File: rewrite-experiment/hypothesis-c/src/risk/exposure-calculator.ts
import { RiskAssessment, RiskStatus } from '../domain/risk';

/**
 * Exposure and VaR calculator, preserving exact formulas from
 * risk-engine ExposureCalculator.java.
 *
 * VaR = notional * volatility * z_score * sqrt(holdingDays / 252)
 *
 * Parameters:
 *   z_score = 2.33 (99% confidence)
 *   holdingDays = 1
 *   tradingDaysPerYear = 252
 *
 * Volatility assumptions:
 *   Equity:    20% (VOL_EQUITY)
 *   FX:         8% (VOL_FX)
 *   Commodity: 30% (VOL_COMMODITY)
 *   Default:   25% (VOL_DEFAULT)
 *
 * VaR flagging threshold: $50,000
 *
 * Exposure:
 *   BUY  -> +notional
 *   SELL -> -notional
 */
export class ExposureCalculator {
  static readonly VOL_EQUITY = 0.20;
  static readonly VOL_FX = 0.08;
  static readonly VOL_COMMODITY = 0.30;
  static readonly VOL_DEFAULT = 0.25;

  static readonly VAR_Z_SCORE = 2.33;
  static readonly HOLDING_PERIOD_DAYS = 1.0;
  static readonly TRADING_DAYS_PER_YEAR = 252.0;
  static readonly VAR_THRESHOLD = 50_000.0;

  static readonly RISK_QUEUE_ORDERS = 'RISK.ORDERS.INBOUND';
  static readonly RISK_QUEUE_RESULTS = 'RISK.RESULTS.OUTBOUND';

  static getVolatility(symbol: string): number {
    if (!symbol) return ExposureCalculator.VOL_DEFAULT;

    if (symbol.includes('/')) return ExposureCalculator.VOL_FX;

    if (['GOLD', 'OIL', 'SILVER'].includes(symbol)) {
      return ExposureCalculator.VOL_COMMODITY;
    }

    return ExposureCalculator.VOL_EQUITY;
  }

  static calculateRisk(assessment: RiskAssessment): void {
    const notional = assessment.quantity * assessment.price;
    assessment.notionalValue = notional;

    if (assessment.side === 'BUY') {
      assessment.exposureContribution = notional;
    } else if (assessment.side === 'SELL') {
      assessment.exposureContribution = -notional;
    } else {
      assessment.exposureContribution = notional;
    }

    const vol = ExposureCalculator.getVolatility(assessment.symbol);
    const sqrtTime = Math.sqrt(
      ExposureCalculator.HOLDING_PERIOD_DAYS / ExposureCalculator.TRADING_DAYS_PER_YEAR
    );
    const varValue = Math.abs(notional) * vol * ExposureCalculator.VAR_Z_SCORE * sqrtTime;
    assessment.varContribution = varValue;

    if (varValue > ExposureCalculator.VAR_THRESHOLD) {
      assessment.riskStatus = RiskStatus.FLAGGED;
    } else {
      assessment.riskStatus = RiskStatus.ASSESSED;
    }
  }
}

File: risk-engine/src/com/bigcorp/risk/engine/ExposureCalculator.java
package com.bigcorp.risk.engine;

import com.bigcorp.risk.model.RiskOrder;

/**
 * Exposure and VaR-style metric calculator.
 * 
 * We compute our own risk metrics rather than using any shared
 * library because the risk team has specific requirements that
 * don't map to the trading engine's model.
 * 
 * VaR is a simplified parametric VaR (delta-normal) using hardcoded
 * volatility constants. Real VaR would need historical simulation
 * but "we'll add that later when we get the data feed." — risk PM
 * 
 * @author contractor (risk team)
 * @since 2017-Q1
 */
public class ExposureCalculator {

    // Hardcoded volatility assumptions per symbol category
    // These should come from a risk config file but the risk PM
    // said "just hardcode them for now, we'll update quarterly"
    private static final double VOL_EQUITY = 0.20;     // 20% annual vol
    private static final double VOL_FX = 0.08;         // 8% for FX
    private static final double VOL_COMMODITY = 0.30;  // 30% for commodities
    private static final double VOL_DEFAULT = 0.25;    // conservative default

    // VaR confidence level (99% = 2.33 standard deviations)
    private static final double VAR_Z_SCORE = 2.33;

    // Time horizon in trading days (1 day VaR)
    private static final double HOLDING_PERIOD_DAYS = 1.0;
    private static final double TRADING_DAYS_PER_YEAR = 252.0;

    // Risk queue name — our own queue, not the shared ones
    public static final String RISK_QUEUE_ORDERS = "RISK.ORDERS.INBOUND";
    public static final String RISK_QUEUE_RESULTS = "RISK.RESULTS.OUTBOUND";

    /**
     * Calculate exposure and VaR for a given risk order.
     * Mutates the RiskOrder with computed values.
     */
    public static void calculateRisk(RiskOrder order) {
        if (order == null) return;

        // Notional value = qty * price
        double notional = order.getQuantity() * order.getPrice();
        order.setNotionalValue(notional);

        // Exposure is directional: BUY = positive, SELL = negative
        double exposure;
        if ("BUY".equals(order.getSide())) {
            exposure = notional;
        } else if ("SELL".equals(order.getSide())) {
            exposure = -notional;
        } else {
            // Unknown side — treat as full exposure
            exposure = notional;
        }
        order.setExposureContribution(exposure);

        // VaR = notional * vol * z-score * sqrt(t/252)
        double vol = getVolatility(order.getSymbol());
        double sqrtTime = Math.sqrt(HOLDING_PERIOD_DAYS / TRADING_DAYS_PER_YEAR);
        double var = Math.abs(notional) * vol * VAR_Z_SCORE * sqrtTime;
        order.setVarContribution(var);

        // Flag if VaR exceeds threshold
        if (var > 50000.0) {
            order.setRiskStatus(RiskOrder.RISK_STATUS_FLAGGED);
        } else {
            order.setRiskStatus(RiskOrder.RISK_STATUS_ASSESSED);
        }
    }

    /**
     * Get assumed volatility for a symbol.
     * Uses hardcoded symbol prefixes because we don't have a proper
     * asset classification service.
     */
    private static double getVolatility(String symbol) {
        if (symbol == null) return VOL_DEFAULT;

        // Very crude classification based on symbol patterns
        // FX pairs contain /
        if (symbol.indexOf('/') >= 0) return VOL_FX;

        // Known commodity symbols
        if ("GOLD".equals(symbol) || "OIL".equals(symbol) || "SILVER".equals(symbol)) {
            return VOL_COMMODITY;
        }

        // Everything else is equity
        return VOL_EQUITY;
    }
}
